src/tests/test_taskrunner.py

```diff
@@ -0,0 +1,123 @@
+            "", # no args
+            "-l", "--list",
+            "-h", "--help", "-?"]
+        for invocation in list_tasks_invocations:
+            response = capture_stdout(lambda: tr.main([invocation]))
+            self.assertEqual(response["result"], SUCCESS_RC)
+
+    def test_commands_are_present(self):
+        "some common commands we always use are in the list and qualified as necessary"
+        expected = [
+            "ssh", # "master.download_keypair", # requires BLDR_ROLE envvar
+            "launch", "masterless.launch",
+            "start", "stop", "restart",
+            "buildvars.switch_revision"
+        ]
+        result = capture_stdout(lambda: tr.main(["--list"]))
+        for case in expected:
+            present = False
+            for row in result["stdout"].splitlines():
+                if row.strip().startswith(case):
+                    present = True
+                    break
+            self.assertTrue(present, "task %r not present in task listing" % case)
+
+    def test_commands_can_be_called(self):
+        task_list = tr.generate_task_list()
+        result_map = tr.exec_task("ping", task_list)
+        expected = {
+            'task': 'ping',
+            'task_args': [],
+            'task_kwargs': {},
+            'result': 'pong',
+            'rc': SUCCESS_RC}
+        self.assertEqual(expected, result_map)
+
+    def test_commands_with_args_can_be_called(self):
+        task_list = tr.generate_task_list()
+        result_map = tr.exec_task("echo:zulu", task_list)
+        expected = {
+            'rc': 0,
+            'result': 'received: zulu',
+            'task': 'echo',
+            'task_args': ['zulu'],
+            'task_kwargs': {}}
+        self.assertEqual(expected, result_map)
+
+    def test_commands_with_args_and_kwargs_can_be_called(self):
+        task_list = tr.generate_task_list()
+        result_map = tr.exec_task("echo:zulu,gamma,a=b", task_list)
+        expected = {
+            'rc': 0,
+            'result': "received: zulu with args: ('gamma',) and kwargs: {'a': 'b'}",
+            'task': 'echo',
+            'task_args': ['zulu', 'gamma'],
+            'task_kwargs': {'a': 'b'}}
+        self.assertEqual(expected, result_map)
+
+    def test_commands_with_whitespace_can_be_called(self):
+        task_list = tr.generate_task_list()
+        # result_map = tr.exec_task("echo:msg,'echo hello world'", task_list) # this is what we give bash
+        result_map = tr.exec_task("echo:msg,echo hello world", task_list) # this is what we see after bash
+        expected = {
+            'task': 'echo',
+            'task_args': ['msg', 'echo hello world'],
+            'task_kwargs': {},
+            'result': "received: msg with args: ('echo hello world',) and kwargs: {}",
+            'rc': SUCCESS_RC}
+        self.assertEqual(expected, result_map)
+
+    def test_commands_with_special_chars_can_be_called_without_quoting(self):
+        task_list = tr.generate_task_list()
```

Comment:
can you add a `this is what we give bash` comment here, too?

Reply:
sure thing